docs/website/docs/extensions/index.md

```diff
@@ -0,0 +1,442 @@
+have several pitfalls but these problems and their solutions are not IREE
+specific and anyone who has designed a shared library interface can apply the
+same rules here in IREE around versioning, performance, etc. One does not add 2
+integers via a syscall and the same holds here: custom modules and the functions
+within should perform a large amount of work to hide overheads involved in the
+cross-module calls and users must be aware that the compiler cannot optimize
+across the call boundaries.
+
+### Pros
+
+* No IREE compiler code changes required.
+* Produced artifacts are portable across IREE deployment configurations.
+* Full system access is allowed - the VM just calls external functions.
+* Runtime modules can be implemented (via shims) in other languages/runtimes.
+
+### Cons
+
+* _(Today)_ no asynchronous scheduling or pipelining.
+  * Future improvements will enable pipelining of host calls.
+* Custom modules must be registered at runtime by the user.
+* The VM custom module ABI goo must be authored by the user (such as with JNI or
+  pybind to move between java/python and C).
+* All custom module code must be compiled and deployed regardless of how much
+  any modules use. The granularity of modules and their versioning is up to the
+  user.
+* Custom module code cannot be optimized by the IREE compiler to avoid
+  host/device readbacks and unnecessary data type conversion.
+
+### When to use
+
+* :heavy_check_mark: Interactions with large libraries or system calls.
+* :heavy_check_mark: Performance-sensitive host code that cannot easily be
+  represented as device code (like UTF-8 string transformation using libicu).
+* :x: _(Today)_ Pipelining or asynchronous execution are required.
+* :x: Extensively using tensor resources.
+
+### Implementation
+
+The current implementation of this requires conversion goo in the compiler to
+properly setup the VM interface. Future improvements will make the compiler
+portion automatic such that in the common case no additional compiler code is
+required beyond inserting the calls into the external modules which can even
+happen before passing the input into the IREE compiler.
+
+The runtime portion requires that the code be exported to the VM system by way
+of an `iree_vm_module_t` interface. A low-level native interface exists with
+minimal overhead and is used for example [by the IREE HAL itself](https://github.com/google/iree/tree/main/iree/modules/hal).
+There is also a C++ wrapper that is significantly easier to work with however it
+needs some performance improvements.
+
+A full end-to-end example can be found under [`iree/samples/custom_modules/`](https://github.com/google/iree/tree/main/iree/samples/custom_modules),
+though it should not currently be considered representative of best practices.
+
+## 3. Extend target-specific device conversion patterns
+
+* TL;DR: add patterns to `iree/Compiler/Codegen/` to emit target code.
```

Comment:
just an idea (which I'm not sold on): you could use an admonition for these TLDRs to get a bit of color/styling

![image](https://user-images.githubusercontent.com/4010439/139908147-aad7fad6-4561-48d7-af2e-9e0cfc478866.png)
```suggestion
## 3. Extend target-specific device conversion patterns

!!! tip
    TL;DR: add patterns to `iree/Compiler/Codegen/` to emit target code.
```

you could also do that for the quote:

![image](https://user-images.githubusercontent.com/4010439/139908785-ddcc3af0-f15b-4f9d-9bdb-e9be9d7870ae.png)